optimization: Moved repeated `-printpriority` fetching out of AddToBlock

l0rinc opened the pull request:
`AddToBlock` was called repeatedly from `addPackageTxs` where the constant value of `printpriority` is recalculated every time.

<img src="https://github.com/bitcoin/bitcoin/assets/1841944/6fd89647-7b6c-4f44-bd04-98d16cd2a938">

This showed up during profiling of AssembleBlock, fetching it once in the constructor results in a small speed increase for many iterations.

> ./src/bench/bench_bitcoin --filter='AssembleBlock' --min-time=10000

before:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          156,460.15 |            6,391.40 |    0.1% |     11.03 | `AssembleBlock`

after:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          149,289.55 |            6,698.39 |    0.3% |     10.97 | `AssembleBlock`

---

The slight speedup shows up in CI as well:
<img src="https://github.com/bitcoin/bitcoin/assets/1841944/3be779c9-2dce-4a96-ae5f-cab5435bd72f">

tdb3 reviewed (approved):
ACK 6e519b83188e77c401bb9024196c499bee67da68

Thanks for optimizing.
Saw around a 6% speedup on my machine.

maflcko reviewed (commented):
Not sure if this is used at all. Maybe it can be removed completely? Otherwise, see the feedback.

paplorinc commented on src/node/miner.h:
```diff
@@ -172,6 +172,7 @@ class BlockAssembler
 
 private:
     const Options m_options;
+    const bool m_fPrintPriority;
```
The [commit](https://github.com/bitcoin/bitcoin/commit/400b15147cf1c4757927935222611e8d3481e739) by @morcos indicates  there's still use for the print statement - even though it wasn't used correctly in the tests, I've removed that reference.

paplorinc commented on test/functional/mining_prioritisetransaction.py:
```diff
@@ -24,10 +24,7 @@
-            "-printpriority=1",
-            "-datacarriersize=100000",
-        ]] * self.num_nodes
+        self.extra_args = [["-datacarriersize=100000"]] * self.num_nodes
```
My bad, I assumed that since nothing was printed it was omitted, but
```
    if (m_print_modified_fee) {
        throw std::runtime_error("m_print_modified_fee is not implemented");
    }
```
fails - thanks for pointing it out.

paplorinc replied:
I've fixed the other Sonarcould warning in a separate commit, if you think it's better that way, I'll merge it with the first one: https://github.com/bitcoin/bitcoin/pull/30324/commits/71753aca8535764f077dc737e92784d2e56f40ef

paplorinc commented on src/node/miner.h:
```diff
@@ -171,6 +172,8 @@ class BlockAssembler
 
 private:
     const Options m_options;
+    const bool m_print_modified_fee = gArgs.GetBoolArg("-printpriority", DEFAULT_PRINTPRIORITY); // -printpriority behavior was changed in 400b15147cf1c4757927935222611e8d3481e739
```
Removed, left it only in the commit message

paplorinc commented on src/node/miner.cpp:
```diff
@@ -407,10 +406,10 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
+            AddToBlock(entry);
             // Erase from the modified set, if present
-            mapModifiedTx.erase(sortedEntries[i]);
+            mapModifiedTx.erase(entry);
```
Makes sense, removed

tdb3 reviewed (approved):
re ACK 7759cd87ccf16e1de8f385924daf4d27afd94464

Tested that the arg still worked properly after the change.  The message was missing without `-printpriority=1`.  The message was present with it.

```
src/bitcoind -debug -regtest -loglevel=trace -printpriority=1
src/bitcoin-cli -regtest createwallet test
src/bitcoin-cli -regtest -generate 101
src/bitcoin-cli -regtest -named sendtoaddress address=bcrt1qcv96frlnhzddst4dnlaa9nvdmvc4darw335wr9 fee_rate=3 amount=2
src/bitcoin-cli -regtest -named sendtoaddress address=bcrt1qnqyyh5xr73lkhfr9tj47u8kevtnahkgrqrnd4s fee_rate=2 amount=6
src/bitcoin-cli getblocktemplate '{"rules": ["segwit"]}'
```

```
2024-06-30T19:43:14Z fee rate 0.00003000 BTC/kvB txid e4bf26ea72e8cbbef9a6e239d1e57fd141c298dbe48696421b753794e20e9922
2024-06-30T19:43:14Z fee rate 0.00002000 BTC/kvB txid 98837b4e4d02399205cbc4e1d21266229b43186fd682e9461a915bd42c1de735
```

Still showing a speedup.  Measured about 5%.

furszy commented on src/node/miner.h:
```diff
@@ -171,6 +172,8 @@ class BlockAssembler
 
 private:
     const Options m_options;
+    const bool m_print_modified_fee = gArgs.GetBoolArg("-printpriority", DEFAULT_PRINTPRIORITY);
```
By doing this, you include `common/args.h` in all files that include `node/miner.h`. It is better to add the field to the `BlockAssembler::Options` struct.

paplorinc replied:
Thanks a lot, it's so much better this way!

l0rinc committed:
Moved the repeated -printpriority fetching out of AddToBlock

AddToBlock was called repeatedly from `addPackageTxs` where the constant value of `printpriority` is recalculated every time.
Since its behavior was changed in 400b151, I've named the variable accordingly.

This showed up during profiling of AssembleBlock, fetching it once in the constructor results in a measurable speed increase for many iterations.

> ./src/bench/bench_bitcoin --filter='AssembleBlock' --min-time=1000

before:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          155,558.97 |            6,428.43 |    0.1% |      1.10 | `AssembleBlock`

after:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          148,083.68 |            6,752.94 |    0.1% |      1.10 | `AssembleBlock`

Co-authored-by: furszy <mfurszy@protonmail.com>

paplorinc commented:
Thanks a lot for your help @tdb3, appreciate your reviews and verifications!

tdb3 reviewed (approved):
re ACK 323ce303086d42292ed9fe7c98e8b459bdf6d1a6

Re-ran same tests in
https://github.com/bitcoin/bitcoin/pull/30324#pullrequestreview-2150128035

same results.

furszy commented on src/node/miner.h:
```diff
@@ -30,7 +30,7 @@ class ChainstateManager;
 
 namespace node {
-static const bool DEFAULT_PRINTPRIORITY = false;
+static const bool DEFAULT_PRINT_MODIFIED_FEE = false;
```
nano nit: would call it "DEFAULT_PRINT_FEERATE" as the "modified" word is internal and might change over time.

paplorinc replied:
I didn't change the external option, just the internal name - as requested here https://github.com/bitcoin/bitcoin/pull/30324#discussion_r1652212085

furszy reviewed (commented):
utACK 323ce303086

achow101 commented:
ACK 323ce303086d42292ed9fe7c98e8b459bdf6d1a6
